### ISSUE TITLE
http/2 server push

### ISSUE THREAD

**Issue by creativeprogramming** [pagespeed/ngx_pagespeed#1086]
Hi,

I'm trying to issue a feature request here, anyway I don't know if it's the right place, and I know that you're probably already working at it if is the case to implement it, but please let me ask you for a comment.

Don't you think that pagespeed has enough  information (inside its 'metadata', after caching optimized resources) to be the 'smart enough' for this? I mean, when I 'dream' http/2 server push in nginx, I think at ngx_pagespeed as the only piece of sotware able to pratically help us implementing it (in a way that will drive us benefits with ease: to dynamically enable this 'optimization' in existing sites)

Maybe the implementation should rely on some underlying (and maybe still missing) nginx feature for the 'transport level' (I don't know, just guessing as they declare that they don't support 'server push' yet), but I guess that the real factor that is preventing nginx guys to implement this, is the lack pratical uses cases for the average nginx user to take advantage of such feature, and not the impossibility to technically spawn a thread to push chunks of data to a client that has just initiated a request. 

Maybe this is out of the targets of this project, i know, but IMO pagespeed is a step forward on this, and can play a great role in helping implementing http/2 server push. It makes sense to think at it, or I'm wrong?
___
Think also at this scenario: if one day nginx  (or a third party) will decide to implement the full server push stack, I mean, also the 'prediction' layer, for instance with an hashmap to cache the resources needed by any URL ([html document URL] --> [array of needed resources...]) what will be better for you after optimizing a webpage? Hacking such 'nginx/third party' internal hashmap to update the URL-->resources metadata? Or to rely on your own implementation? Maybe the first one, ok, but I mean, probably one day pagespeed will face the issue to correctly interact with some server push implementations, in apache or nginx, so maybe it makes sense to start thinking at its role on this.

**Comment by jeffkaufman** [comment]
Yes, this is something we want to do at some point.  We have a lot of information about which resources are used, and once there are servers that will let us push we should start using it.
